qutebrowser/misc/sql.py

```diff
@@ -168,82 +171,212 @@ def raise_sqlite_error(msg, error):
+                f"{self._user_version}, but {self._USER_VERSION.major}.x is supported)")
+
+        if self.user_version_changed():
+            # Enable write-ahead-logging and reduce disk write frequency
+            # see https://sqlite.org/pragma.html and issues #2930 and #3507
+            #
+            # We might already have done this (without a migration) in earlier versions,
+            # but as those are idempotent, let's make sure we run them once again.
+            self.query("PRAGMA journal_mode=WAL").run()
+            self.query("PRAGMA synchronous=NORMAL").run()
+
+    def qSqlDatabase(self) -> QSqlDatabase:
+        """Return the wrapped QSqlDatabase instance."""
+        database = QSqlDatabase.database(self._path, open=True)
+        if not database.isValid():
+            raise BugError('Failed to get connection. Did you close() this Database '
+                           'instance?')
+        return database
+
+    def query(self, querystr: str, forward_only: bool = True) -> 'Query':
+        """Return a Query instance linked to this Database."""
+        return Query(self, querystr, forward_only)
+
+    def table(self, name: str, fields: List[str],
+              constraints: Optional[Dict[str, str]] = None,
+              parent: Optional[QObject] = None) -> 'SqlTable':
+        """Return a SqlTable instance linked to this Database."""
+        return SqlTable(self, name, fields, constraints, parent)
+
+    def user_version_changed(self) -> bool:
+        """Whether the version stored in the database differs from the current one."""
+        return self._user_version != self._USER_VERSION
+
+    def upgrade_user_version(self) -> None:
+        """Upgrade the user version to the latest version.
+
+        This method should be called once all required operations to migrate from one
+        version to another have been run.
+        """
+        log.sql.debug(f"Migrating from version {self._user_version} "
+                      f"to {self._USER_VERSION}")
+        self.query(f'PRAGMA user_version = {self._USER_VERSION.to_int()}').run()
+        self._user_version = self._USER_VERSION
+
+    def close(self) -> None:
+        """Close the SQL connection."""
+        database = self.qSqlDatabase()
+        database.close()
+        sip.delete(database)
+        QSqlDatabase.removeDatabase(self._path)
+
+    def transaction(self) -> 'Transaction':
+        """Return a Transaction object linked to this Database."""
+        return Transaction(self)
+
+    @classmethod
+    def version(cls) -> str:
+        """Return the sqlite version string."""
+        try:
+            in_memory_db = cls(':memory:')
+            version = in_memory_db.query("select sqlite_version()").run().value()
+            in_memory_db.close()
+            return version
+        except KnownError as e:
+            return f'UNAVAILABLE ({e})'
+
+
+class Transaction(contextlib.AbstractContextManager[None]):
```

Comment:
This only works in Python 3.9+ (we still support 3.6, 3.7 and 3.8 for a while). I suppose we can leave the `[None]` off for now.

Reply:
Ah, I only added this because mypy was complaining. Does mypy run with Python 3.9 in the Github workflows?

Reply:
I don't think the Python version mypy runs with is relevant - its checks are based on the `python_version` set in `.mypy.ini`, which is set to 3.6 (the oldest version we support). However, I expect this to fail with older Python versions at runtime:

```pycon
Python 3.6.13 (default, Feb 25 2021, 14:10:08) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import contextlib
>>> contextlib.AbstractContextManager[None]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'ABCMeta' object is not subscriptable
```

Though it doesn't seem to fail immediately, not really sure why!

As for mypy, it does indeed complain due to the `disallow_any_generics = True` in the config. I'd suggest

```suggestion
class Transaction(contextlib.AbstractContextManager):  # type: ignore[type-arg]
```

until we drop support for older Pythons, we have a couple of other places doing this already.